asteroid/separate.py

```diff
@@ -120,50 +121,84 @@ def file_separate(
+    save_name_template = base + "_est{}.wav"
+
+    # Bail out early if an estimate file already exists and we shall not overwrite.
+    est1_filename = save_name_template.format(1)
+    if os.path.isfile(est1_filename) and not force_overwrite:
+        warnings.warn(
+            f"File {est1_filename} already exists, pass `force_overwrite=True` to overwrite it",
+            UserWarning,
+        )
+        return
+
     # SoundFile wav shape: [time, n_chan]
-    wav, fs = sf.read(filename, dtype="float32", always_2d=True)
+    wav, fs = _load_audio(filename)
     if wav.shape[-1] > 1:
         warnings.warn(
             f"Received multichannel signal with {wav.shape[-1]} signals, "
             f"using the first channel only."
         )
     # FIXME: support only single-channel files for now.
-    if fs != model.sample_rate:
-        if resample:
-            from librosa import resample
-
-            wav = resample(wav[:, 0], orig_sr=fs, target_sr=model.sample_rate)[:, None]
-        else:
-            raise RuntimeError(
-                f"Received a signal with a sampling rate of {fs}Hz for a model "
-                f"of {model.sample_rate}Hz. You can pass `resample=True` to resample automatically."
-            )
+    if resample:
+        wav = _resample(wav[:, 0], orig_sr=fs, target_sr=int(model.sample_rate))[:, None]
+    elif fs != model.sample_rate:
+        raise RuntimeError(
+            f"Received a signal with a sampling rate of {fs}Hz for a model "
+            f"of {model.sample_rate}Hz. You can pass `resample=True` to resample automatically."
+        )
     # Pass wav as [batch, n_chan, time]; here: [1, 1, time]
     wav = wav[:, 0][None, None]
-    (to_save,) = numpy_separate(model, wav, **kwargs)
+    (est_srcs,) = numpy_separate(model, wav, **kwargs)
+    # Resample to original sr
+    est_srcs = [
+        _resample(est_src, orig_sr=int(model.sample_rate), target_sr=fs) for est_src in est_srcs
+    ]
 
     # Save wav files to filename_est1.wav etc...
-    for src_idx, est_src in enumerate(to_save):
-        base = ".".join(filename.split(".")[:-1])
-        save_name = base + "_est{}.".format(src_idx + 1) + filename.split(".")[-1]
-        if output_dir is not None:
-            save_name = os.path.join(output_dir, save_name.split("/")[-1])
-        if os.path.isfile(save_name) and not force_overwrite:
-            warnings.warn(
-                f"File {save_name} already exists, pass `force_overwrite=True` to overwrite it",
-                UserWarning,
-            )
-            return
-        if fs != model.sample_rate:
-            from librosa import resample
-
-            est_src = resample(est_src, orig_sr=model.sample_rate, target_sr=fs)
-        sf.write(save_name, est_src, fs)
+    for src_idx, est_src in enumerate(est_srcs, 1):
+        sf.write(save_name_template.format(src_idx), est_src, fs)
+
+
+def _resample(wav, orig_sr, target_sr, _resamplers={}):
```

Comment:
This is cool, and I've never seen it. 
I wonder if it's not too intricate though.

Reply:
You mean the caching pattern? I can make it more explicit if you want. Maybe adding "cache" in the name would be helpful already.

Reply:
> You mean the caching pattern?

Yes. 

Including it in the name is good. `_autocached_resamplers={}` as a kwarg? 
My first idea was that I'd have to pass it to benefit from the caching and then I understood the trick because it's mutable. 

By the way, do you know where the `_resamplers` object lives? A link to the logic would be cool.

Reply:
You mean the dictionary? It "lives in" (to be exact, is referenced by) the `_resample` function object. Funny how I take this caching pattern for granted. I don't actually know if its a wide spread pattern. I think before adding more documentation on how it works we should rather move it from the function object to the module, which is more obvious to most people.

Reply:
It's just my Python level which is not too high. 
If it's in the module, it's much simpler to get indeed.
 
For module-level `_resamplers`, it's in `globals` and it's modified from there, that seems simple. But, when it's in the kwargs default, I don't yet understand what's happening under the hood. It's not in `globals` for sure, where is the reference?

Reply:
I'm gonna read about it :smiley:

Reply:
https://docs.python.org/3/faq/programming.html#why-are-default-values-shared-between-objects

Sounds like it's a common pattern. 
Happy that I learned it!

Reply:
Oh indeed! Nice find!

> It's not in `globals` for sure, where is the reference?

`globals()` actually returns the dict (= attributes) of the current module object (note everything is an object in Python, even ints and modules and functions). In the module attributes you find the function object. When you do `sf.read(...)` the Python interpreter looks for the `read` attribute of the `sf` (module) object and then calls it. `read` itself is an object (a function object) that contains attributes, like the actual byte code that defines the function, the list of type annotations in the signature, ... and the default values of all arguments:


```pysh
>>> def foo(a: dict = {}):
...  x = a["somekey"]
...  return x
...
>>> foo.__code__
<code object foo at 0x105435710, file "<stdin>", line 1>
>>> foo.__code__.co_argcount
1
>>> foo.__code__.co_nlocals
2
>>> foo.__code__.co_varnames
('a', 'x')
>>> foo.__annotations__
{'a': <class 'dict'>}

>>> foo.__defaults__  # argument default values
({},)
>>> foo()  # This will obviously fail because the default dict is empty
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in foo
KeyError: 'somekey'
>>> foo.__defaults__[0]["somekey"] = 42   # We can modify the default values (never do that kind of f*ckery)
>>> foo()
42
```

Reply:
Ok, so it's defined in `foo.__default__` when the function is defined and never updated, that's perfect ! 
Thanks for the detailed answer.